src/sage/rings/completion_polynomial_ring.py

```diff
@@ -0,0 +1,858 @@
+        backend = self.backend(force=True)
+        if backend.valuation() < 0:
+            raise ValueError("this element of negative valuation cannot be approximated by a polynomial")
+        try:
+            backend = backend.power_series()
+        except AttributeError:
+            pass
+        d = p.degree()
+        k = S.residue_field()
+        if prec is not None:
+            backend = backend.add_bigoh(prec)
+        prec = backend.precision_absolute()
+        f = backend.polynomial().change_ring(k)
+        g = f(f.parent().gen() - k.gen())
+        gs = [A([c[i] for c in g.list()]) for i in range(d)]
+        if prec is Infinity:
+            for i in range(1, d):
+                if gs[i]:
+                    raise ValueError("this element is exact and does not lie in the polynomial ring")
+            return gs[0]
+        xbar = S._xbar(prec)
+        modulus = p ** prec
+        res = gs[0]
+        xbari = A.one()
+        for i in range(1, d):
+            xbari = (xbari * xbar) % modulus
+            res += xbari * gs[i]
+        return res % modulus
+
+    def shift(self, n):
+        raise NotImplementedError
+
+
+class CompletionPolynomialRing(UniqueRepresentation, RingExtension_generic):
+    r"""
+    A class for completions of polynomial rings and their fraction fields.
+    """
+    Element = CompletionPolynomial
+
+    def __classcall_private__(cls, ring, p, default_prec=20, sparse=False):
+        r"""
+        Normalize the parameters and call the appropriate constructor.
+
+        INPUT:
+
+        - ``ring`` -- the underlying polynomial ring or field
+
+        - ``p`` -- a generator of the ideal at which the completion is
+          done; it could also be ``Infinity``
+
+        - ``default_pres`` (default: ``20``) -- the default precision
+          of the completion
+
+        - ``sparse`` (default: ``False``) -- a boolean
+
+        TESTS::
+
+            sage: A.<x> = ZZ[]
+            sage: A.completion(x^2 + x + 1)
+            Completion of Univariate Polynomial Ring in x over Integer Ring at x^2 + x + 1
+            sage: A.completion(2*x - 1)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: the leading coefficient of p must be invertible
+
+        ::
+
+            sage: B.<y> = QQ[]
+            sage: B.completion(x^2 + x^3)
+            Traceback (most recent call last):
+            ...
+            ValueError: p must be a squarefree polynomial
+
+        When passing a variable name, a standard ring of power series is
+        constructed::
+
+            sage: A.completion('x')
+            Power Series Ring in x over Integer Ring
+            sage: A.completion(x)
+            Completion of Univariate Polynomial Ring in x over Integer Ring at x
+        """
+        if isinstance(ring, PolynomialRing_generic):
+            if isinstance(p, str):
```

Comment:
```suggestion
            if isinstance(p, str) or p == ring.gen():
```
As mentioned on another comment, I think it is good to special case this.

Reply:
On the other hand, I think it is important to have something uniform (with the same parent, the same class) when the ideal varies.
It could be a mess if you do
```
p = <complicated calculation>
Ap = A.completion(p)
```
and the result you get depends on the fact that the result of the complicated calculation is or is not the generator. 
Another issue concerns inexact base rings when any approximation of a generator, in particular `O(1)*x` will be recognized a generator.

Reply:
Then power series rings should be subclasses. The point is that there we can take advantage of the specific input, so we can and should optimize it (and have other things like natural coercions). Although as long as the API is the same, ducktyping will make it okay, but this would be a code smell to me. Also the output type depends on the precision, so there might be a bigger refactoring needed...

Reply:
I do not really agree.
According to me, it makes sense that `CompletionRing` and `PowerSeriesRing` are two different types of objects even though the former relies on the latter for the implementation.
If the user wants to create a `PowerSeriesRing`, then he/she can still do it as before.
If he/she really wants efficiency, he/she can always call the method `power_series_ring` to get the backend.

Reply:
Can you elaborate on them being two different types of objects? It seems more natural to me that FPS are just a special type of completion, so I would have them have a subclass relation. Does it have to do with the generality of the base ring for a FPS versus a completion being more limited? Sorry if I’m missing something obvious.

Reply:
> If the user wants to create a `PowerSeriesRing`, then he/she can still do it as before.
> If he/she really wants efficiency, he/she can always call the method `power_series_ring` to get the backend.

I think this is the opposite direction. @tscrim is suggesting making `PowerSeriesRing` subclass of `CompletionRing`, not vice versa.

`PowerSeriesRing` is indeed a special case of `CompletionRing` (both in interface and in implementation). Although _in practice_ the old interface (of `PowerSeriesRing`) may have some issues that make it incompatible with any general interface, e.g. some functions returning `"x"` instead of `x`.